samples/caf_sensor_manager/src/modules/sensor_stub_gen.c

```diff
@@ -0,0 +1,172 @@
+ */
+
+#include <zephyr/kernel.h>
+#include <drivers/sensor_stub.h>
+
+#define MODULE sensor_stub_gen
+#include <caf/events/module_state_event.h>
+
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_SENSOR_STUB_GEN_LOG_LEVEL);
+
+
+#define SENSOR_FRACTIONAL_MAX (1000000)
+#define SENSOR_STUB_GEN_X_STEP (0.1)
+#define SENSOR_STUB_GEN_X_MAX  (5.0)
+#define SENSOR_STUB_GEN_X_MIN  (-5.0)
+
+#define SENSOR_STUB_GEN_Y_STEP (0.1)
+#define SENSOR_STUB_GEN_Y_MAX  (2.0)
+#define SENSOR_STUB_GEN_Y_MIN  (-2.0)
+
+#define SENSOR_STUB_GEN_Z_STEP (0.2)
+#define SENSOR_STUB_GEN_Z_MAX  (1.0)
+#define SENSOR_STUB_GEN_Z_MIN  (0.0)
+
+#define SENSOR_STUB_SENSOR_FP(x) ((int64_t)((x) * SENSOR_FRACTIONAL_MAX))
+
+enum {
+	SENSOR_STUB_CH_X,
+	SENSOR_STUB_CH_Y,
+	SENSOR_STUB_CH_Z,
+	SENSOR_STUB_CH_CNT,
+};
+
+/* All the data for implemented sensor simulation */
+struct sensor_stub_gen_ch_data {
+	/* Current value */
+	struct sensor_value val;
+	/* Current direction of the changes */
+	bool dir;
+};
+
+/* Parameters of the signal generated on the selected channel */
+struct sensor_stub_gen_ch_params {
+	int64_t step, max, min;
+};
+
+/* SENSOR_CHAN_ACCEL_X, SENSOR_CHAN_ACCEL_Y, SENSOR_CHAN_ACCEL_Z */
+static struct sensor_stub_gen_ch_data gen_data[SENSOR_STUB_CH_CNT];
+
+static const struct sensor_stub_gen_ch_params gen_params[SENSOR_STUB_CH_CNT] = {
+	[SENSOR_STUB_CH_X] = {
+		.step = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_X_STEP),
+		.min  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_X_MIN),
+		.max  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_X_MAX)
+	},
+	[SENSOR_STUB_CH_Y] = {
+		.step = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Y_STEP),
+		.min  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Y_MIN),
+		.max  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Y_MAX)
+	},
+	[SENSOR_STUB_CH_Z] = {
+		.step = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Z_STEP),
+		.min  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Z_MIN),
+		.max  = SENSOR_STUB_SENSOR_FP(SENSOR_STUB_GEN_Z_MAX)
+	}
+};
+
+static inline int64_t sensor_val_2_fp(const struct sensor_value *val)
+{
+	return ((int64_t)val->val1 * SENSOR_FRACTIONAL_MAX) + val->val2;
+}
+
+static inline void fp_2_sensor_val(struct sensor_value *out, int64_t val)
+{
+	out->val1 = (int32_t)(val / SENSOR_FRACTIONAL_MAX);
+	out->val2 = (int32_t)(val % SENSOR_FRACTIONAL_MAX);
+}
+
+/* Generate changes on selected channel with given parameters */
+static inline void sensor_stub_gen_chan(struct sensor_stub_gen_ch_data *ch_data,
+					const struct sensor_stub_gen_ch_params *params)
+{
+	int64_t val = sensor_val_2_fp(&ch_data->val);
+
+	if (ch_data->dir) {
+		val += params->step;
+		if (val >= params->max) {
+			ch_data->dir = false;
+		}
+	} else {
+		val -= params->step;
+		if (val <= params->min) {
+			ch_data->dir = true;
+		}
+	}
+
+	fp_2_sensor_val(&ch_data->val, val);
+}
+
+int sensor_stub_gen_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ALL:
```

Comment:
Fetch supports `SENSOR_CHAN_ALL` but get does not, why is that?

Reply:
Because you cannot get all the channels at once. This way it also works in real drivers - see temp_nrf5_mpsl - you can fetch ALL, but you cannot ask for ALL data. The same for bh1749, paw3212, sensor_sim.... Just all the drivers I had quickly look at.

Reply:
gosh... okay.